File: accounting_reports/static/src/js/tax_report.js
/** @odoo-module **/
import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { _t } from "@web/core/l10n/translation";

const { Component, useState, onWillStart } = owl;

export class TaxReport extends Component {
    static template = "accounting_reports.TaxReport";
    static props = ["*"];

    setup() {
        this.actionService = useService("action");
        this.orm = useService("orm");

        this.state = useState({
            // Report Data
            reportData: { 
                sales_taxes: [], 
                purchase_taxes: [], 
                totals: {} 
            },

            // Expanded states
            expandedSalesTaxes: {},
            expandedPurchaseTaxes: {},
            taxDetails: {},

            // Filters
            dateFrom: this.getFirstDayOfYear(),
            dateTo: this.getTodayDate(),
            postedOnly: true,

            // UI
            isLoading: true,
            loadingDetails: {},
        });

        onWillStart(async () => {
            await this.loadReportData();
        });
    }

    // ==================== DATE HELPERS ====================

    getFirstDayOfYear() {
        const now = new Date();
        return `${now.getFullYear()}-01-01`;
    }

    getTodayDate() {
        const now = new Date();
        return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')}`;
    }

    getFirstDayOfMonth() {
        const now = new Date();
        return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-01`;
    }

    getFirstDayOfQuarter() {
        const now = new Date();
        const quarter = Math.floor(now.getMonth() / 3);
        const month = quarter * 3 + 1;
        return `${now.getFullYear()}-${String(month).padStart(2, '0')}-01`;
    }

    // ==================== DATA LOADING ====================

    async loadReportData() {
        this.state.isLoading = true;
        try {
            const data = await this.orm.call(
                'report.tax.report',
                'get_report_data',
                [],
                {
                    date_from: this.state.dateFrom,
                    date_to: this.state.dateTo,
                    posted_only: this.state.postedOnly,
                }
            );

            this.state.reportData = data;
            // Reset expanded states
            this.state.expandedSalesTaxes = {};
            this.state.expandedPurchaseTaxes = {};
            this.state.taxDetails = {};

        } catch (error) {
            console.error("Error loading report data:", error);
        }
        this.state.isLoading = false;
    }

    async loadTaxDetails(taxId, taxType) {
        const key = `${taxType}_${taxId}`;
        
        if (this.state.taxDetails[key]) {
            return; // Already loaded
        }

        this.state.loadingDetails[key] = true;
        
        try {
            const details = await this.orm.call(
                'report.tax.report',
                'get_tax_details',
                [],
                {
                    tax_id: taxId,
                    date_from: this.state.dateFrom,
                    date_to: this.state.dateTo,
                    tax_type: taxType,
                    posted_only: this.state.postedOnly,
                }
            );

            this.state.taxDetails[key] = details;
        } catch (error) {
            console.error("Error loading tax details:", error);
        }
        
        this.state.loadingDetails[key] = false;
    }

    // ==================== FORMATTING ====================

    formatNumber(num) {
        return (num || 0).toLocaleString('en-US', {
            minimumFractionDigits: 2,
            maximumFractionDigits: 2
        });
    }

    formatPercent(num) {
        return (num || 0).toFixed(2) + '%';
    }

    // ==================== EXPAND/COLLAPSE ====================

    async toggleSalesTax(taxId) {
        const isExpanded = this.state.expandedSalesTaxes[taxId];
        this.state.expandedSalesTaxes[taxId] = !isExpanded;
        
        if (!isExpanded) {
            await this.loadTaxDetails(taxId, 'sale');
        }
    }

    async togglePurchaseTax(taxId) {
        const isExpanded = this.state.expandedPurchaseTaxes[taxId];
        this.state.expandedPurchaseTaxes[taxId] = !isExpanded;
        
        if (!isExpanded) {
            await this.loadTaxDetails(taxId, 'purchase');
        }
    }

    isSalesTaxExpanded(taxId) {
        return this.state.expandedSalesTaxes[taxId] || false;
    }

    isPurchaseTaxExpanded(taxId) {
        return this.state.expandedPurchaseTaxes[taxId] || false;
    }

    getTaxDetails(taxId, taxType) {
        const key = `${taxType}_${taxId}`;
        return this.state.taxDetails[key] || [];
    }

    isLoadingDetails(taxId, taxType) {
        const key = `${taxType}_${taxId}`;
        return this.state.loadingDetails[key] || false;
    }

    expandAllSales() {
        for (const tax of this.state.reportData.sales_taxes) {
            this.state.expandedSalesTaxes[tax.tax_id] = true;
            this.loadTaxDetails(tax.tax_id, 'sale');
        }
    }

    expandAllPurchases() {
        for (const tax of this.state.reportData.purchase_taxes) {
            this.state.expandedPurchaseTaxes[tax.tax_id] = true;
            this.loadTaxDetails(tax.tax_id, 'purchase');
        }
    }

    collapseAll() {
        this.state.expandedSalesTaxes = {};
        this.state.expandedPurchaseTaxes = {};
    }

    // ==================== FILTER HANDLERS ====================

    async onChangeDateFrom(ev) {
        this.state.dateFrom = ev.target.value;
        await this.loadReportData();
    }

    async onChangeDateTo(ev) {
        this.state.dateTo = ev.target.value;
        await this.loadReportData();
    }

    async onChangePostedOnly(ev) {
        this.state.postedOnly = ev.target.checked;
        await this.loadReportData();
    }

    async setPeriod(period) {
        const now = new Date();
        const year = now.getFullYear();
        const month = now.getMonth();

        switch(period) {
            case 'this_month':
                this.state.dateFrom = `${year}-${String(month + 1).padStart(2, '0')}-01`;
                this.state.dateTo = this.getTodayDate();
                break;
            case 'last_month':
                const lastMonth = month === 0 ? 12 : month;
                const lastMonthYear = month === 0 ? year - 1 : year;
                const lastDay = new Date(lastMonthYear, lastMonth, 0).getDate();
                this.state.dateFrom = `${lastMonthYear}-${String(lastMonth).padStart(2, '0')}-01`;
                this.state.dateTo = `${lastMonthYear}-${String(lastMonth).padStart(2, '0')}-${lastDay}`;
                break;
            case 'this_quarter':
                const quarter = Math.floor(month / 3);
                const qMonth = quarter * 3 + 1;
                this.state.dateFrom = `${year}-${String(qMonth).padStart(2, '0')}-01`;
                this.state.dateTo = this.getTodayDate();
                break;
            case 'this_year':
                this.state.dateFrom = `${year}-01-01`;
                this.state.dateTo = this.getTodayDate();
                break;
        }
        await this.loadReportData();
    }

    // ==================== ACTIONS ====================

    async onRefresh() {
        await this.loadReportData();
    }

    onExportExcel() {
        let csv = 'Type,Tax Name,Rate,Base Amount,Tax Amount,Invoices\n';

        // Sales Taxes
        csv += '\nSALES TAXES\n';
        for (const tax of this.state.reportData.sales_taxes) {
            csv += `"Sales","${tax.tax_name}",${tax.tax_percent}%,${tax.base_amount},${tax.tax_total},${tax.invoice_count}\n`;
        }
        csv += `"","TOTAL SALES","",${this.state.reportData.totals.sales_base},${this.state.reportData.totals.sales_tax},""\n`;

        // Purchase Taxes
        csv += '\nPURCHASE TAXES\n';
        for (const tax of this.state.reportData.purchase_taxes) {
            csv += `"Purchases","${tax.tax_name}",${tax.tax_percent}%,${tax.base_amount},${tax.tax_total},${tax.invoice_count}\n`;
        }
        csv += `"","TOTAL PURCHASES","",${this.state.reportData.totals.purchase_base},${this.state.reportData.totals.purchase_tax},""\n`;

        // Net
        csv += `\n"","NET TAX DUE","","",${this.state.reportData.totals.net_tax},""\n`;

        const BOM = '\uFEFF';
        const blob = new Blob([BOM + csv], { type: 'text/csv;charset=utf-8;' });
        const link = document.createElement('a');
        link.href = URL.createObjectURL(blob);
        link.download = `tax_report_${this.state.dateFrom}_to_${this.state.dateTo}.csv`;
        link.click();
    }

    onPrint() {
        window.print();
    }

    // ==================== NAVIGATION ====================

    openMove(moveId) {
        this.actionService.doAction({
            type: "ir.actions.act_window",
            name: _t("Journal Entry"),
            res_model: "account.move",
            res_id: moveId,
            views: [[false, "form"]],
            target: "current",
        });
    }

    openTaxMoves(taxData, taxType) {
        const moveIds = taxData.move_ids || [];
        if (moveIds.length === 0) return;

        this.actionService.doAction({
            type: "ir.actions.act_window",
            name: taxData.tax_name,
            res_model: "account.move",
            views: [[false, "list"], [false, "form"]],
            domain: [['id', 'in', moveIds]],
            target: "current",
        });
    }

    // ==================== HELPERS ====================

    getMoveTypeLabel(moveType) {
        const labels = {
            'out_invoice': 'Invoice',
            'out_refund': 'Credit Note',
            'in_invoice': 'Bill',
            'in_refund': 'Refund',
            'entry': 'Journal Entry',
        };
        return labels[moveType] || moveType;
    }

    getNetTaxClass() {
        const net = this.state.reportData.totals?.net_tax || 0;
        if (net > 0) return 'net-positive';  // You owe tax
        if (net < 0) return 'net-negative';  // Tax refund
        return '';
    }
}

TaxReport.template = "accounting_reports.TaxReport";
registry.category("actions").add("tax_report", TaxReport);

File: accounting_reports/models/tax_report.py
# -*- coding: utf-8 -*-
from odoo import api, fields, models
from odoo.tools import float_is_zero
from datetime import datetime


class TaxReport(models.AbstractModel):
    _name = 'report.tax.report'
    _description = 'Tax Report'

    @api.model
    def get_report_data(self, date_from=False, date_to=False, posted_only=True):
        """
        Get Tax report data grouped by tax.
        
        Returns:
        {
            'sales_taxes': [
                {
                    'tax_id': 1,
                    'tax_name': 'VAT 5% (Sales)',
                    'tax_amount': 5.0,
                    'base_amount': 100000.00,
                    'tax_total': 5000.00,
                    'invoice_count': 25,
                }
            ],
            'purchase_taxes': [...],
            'totals': {
                'sales_base': 100000.00,
                'sales_tax': 5000.00,
                'purchase_base': 80000.00,
                'purchase_tax': 4000.00,
                'net_tax': 1000.00,  # Sales Tax - Purchase Tax
            }
        }
        """
        MoveLine = self.env['account.move.line']
        
        # Build domain
        domain = [
            ('tax_line_id', '!=', False),  # Only tax lines
            ('parent_state', '=', 'posted') if posted_only else ('parent_state', '!=', 'cancel'),
        ]
        
        if date_from:
            if isinstance(date_from, str):
                date_from = datetime.strptime(date_from, '%Y-%m-%d').date()
            domain.append(('date', '>=', date_from))
        
        if date_to:
            if isinstance(date_to, str):
                date_to = datetime.strptime(date_to, '%Y-%m-%d').date()
            domain.append(('date', '<=', date_to))
        
        # Get all tax lines
        tax_lines = MoveLine.search(domain)
        
        # Group by tax
        sales_taxes = {}
        purchase_taxes = {}
        
        for line in tax_lines:
            tax = line.tax_line_id
            tax_id = tax.id
            
            # Get base amount from the tax base
            # tax_base_amount is the base on which this tax was computed
            base_amount = abs(line.tax_base_amount)
            tax_amount = abs(line.balance)
            
            # Determine if sales or purchase tax
            move_type = line.move_id.move_type
            
            if move_type in ('out_invoice', 'out_refund', 'out_receipt'):
                # Sales tax
                tax_dict = sales_taxes
                # For refunds, amounts might be negative
                if move_type == 'out_refund':
                    base_amount = -base_amount
                    tax_amount = -tax_amount
            elif move_type in ('in_invoice', 'in_refund', 'in_receipt'):
                # Purchase tax
                tax_dict = purchase_taxes
                # For refunds, amounts might be negative
                if move_type == 'in_refund':
                    base_amount = -base_amount
                    tax_amount = -tax_amount
            else:
                # Journal entries - check tax type
                if tax.type_tax_use == 'sale':
                    tax_dict = sales_taxes
                elif tax.type_tax_use == 'purchase':
                    tax_dict = purchase_taxes
                else:
                    continue
            
            if tax_id not in tax_dict:
                tax_dict[tax_id] = {
                    'tax_id': tax_id,
                    'tax_name': tax.name,
                    'tax_percent': tax.amount,
                    'tax_type': tax.amount_type,
                    'base_amount': 0.0,
                    'tax_total': 0.0,
                    'invoice_count': 0,
                    'move_ids': set(),
                }
            
            tax_dict[tax_id]['base_amount'] += base_amount
            tax_dict[tax_id]['tax_total'] += tax_amount
            tax_dict[tax_id]['move_ids'].add(line.move_id.id)
        
        # Convert to list and calculate invoice count
        sales_list = []
        for tax_data in sorted(sales_taxes.values(), key=lambda x: x['tax_name']):
            tax_data['invoice_count'] = len(tax_data['move_ids'])
            tax_data['move_ids'] = list(tax_data['move_ids'])
            sales_list.append(tax_data)
        
        purchase_list = []
        for tax_data in sorted(purchase_taxes.values(), key=lambda x: x['tax_name']):
            tax_data['invoice_count'] = len(tax_data['move_ids'])
            tax_data['move_ids'] = list(tax_data['move_ids'])
            purchase_list.append(tax_data)
        
        # Calculate totals
        sales_base = sum(t['base_amount'] for t in sales_list)
        sales_tax = sum(t['tax_total'] for t in sales_list)
        purchase_base = sum(t['base_amount'] for t in purchase_list)
        purchase_tax = sum(t['tax_total'] for t in purchase_list)
        
        return {
            'sales_taxes': sales_list,
            'purchase_taxes': purchase_list,
            'totals': {
                'sales_base': sales_base,
                'sales_tax': sales_tax,
                'purchase_base': purchase_base,
                'purchase_tax': purchase_tax,
                'net_tax': sales_tax - purchase_tax,
            },
            'date_from': date_from.strftime('%d/%m/%Y') if date_from else '',
            'date_to': date_to.strftime('%d/%m/%Y') if date_to else '',
        }

    @api.model
    def get_tax_details(self, tax_id, date_from=False, date_to=False, 
                        tax_type='sale', posted_only=True):
        """
        Get detailed moves for a specific tax.
        
        Returns list of invoices/moves with this tax.
        """
        MoveLine = self.env['account.move.line']
        
        domain = [
            ('tax_line_id', '=', tax_id),
            ('parent_state', '=', 'posted') if posted_only else ('parent_state', '!=', 'cancel'),
        ]
        
        if date_from:
            if isinstance(date_from, str):
                date_from = datetime.strptime(date_from, '%Y-%m-%d').date()
            domain.append(('date', '>=', date_from))
        
        if date_to:
            if isinstance(date_to, str):
                date_to = datetime.strptime(date_to, '%Y-%m-%d').date()
            domain.append(('date', '<=', date_to))
        
        # Filter by move type based on tax_type
        if tax_type == 'sale':
            domain.append(('move_id.move_type', 'in', ('out_invoice', 'out_refund', 'out_receipt', 'entry')))
        else:
            domain.append(('move_id.move_type', 'in', ('in_invoice', 'in_refund', 'in_receipt', 'entry')))
        
        tax_lines = MoveLine.search(domain, order='date desc')
        
        # Group by move
        moves_data = {}
        for line in tax_lines:
            move = line.move_id
            move_id = move.id
            
            if move_id not in moves_data:
                moves_data[move_id] = {
                    'move_id': move_id,
                    'move_name': move.name,
                    'date': line.date.strftime('%d/%m/%Y') if line.date else '',
                    'partner_name': move.partner_id.name or '',
                    'move_type': move.move_type,
                    'base_amount': 0.0,
                    'tax_amount': 0.0,
                }
            
            base = line.tax_base_amount
            tax = line.balance
            
            # Handle refunds
            if move.move_type in ('out_refund', 'in_refund'):
                base = -abs(base)
                tax = -abs(tax)
            else:
                base = abs(base)
                tax = abs(tax)
            
            moves_data[move_id]['base_amount'] += base
            moves_data[move_id]['tax_amount'] += tax
        
        return list(moves_data.values())
